build fails on assemble - "application not found on play  (404)"

### Describe the bug

When calling gradlew assemble from buildscript with multiple flavors, the build fails in ':app:processDevReleaseMetadata', due to a call to https://www.googleapis.com/androidpublisher/v3/applications/>flavor-package</edits, which fails on 404 because this flavor isnt in play.

Is this intended behavior or a bug, and is there a way to configure my way out of this problem?

Thanks in advance :)

### To Reproduce

run assemble on a project with diffferent flavors which isnt published on play.

#### Versions

 - gralde-play-pub:2.0.0-beta1
 - gradle wrapper: 4.6
 - android plugin: 3.2.1

### Expected behavior

I dont expect your plugin to run on tasks which arent release-related.

### Additional context

POST https://www.googleapis.com/androidpublisher/v3/applications/com....../edits
Accept-Encoding: gzip
Authorization: <Not Logged>
User-Agent: gradle-play-publisher Google-API-Java-Client Google-HTTP-Java-Client/1.23.0 (gzip)
Content-Encoding: gzip
Content-Length: 20

15:11:25.069 [INFO] [com.google.api.client.http.HttpTransport] curl -v --compressed -X POST -H 'Accept-Encoding: gzip' -H 'Authorization: <Not Logged>' -H 'User-Agent: gradle-play-publisher Google-API-Java-Client Google-HTTP-Java-Client/1.23.0 (gzip)' -H 'Content-Encoding: gzip' -d '@-' -- 'https://www.googleapis.com/androidpublisher/v3/applications/c
om..../edits' << $$$
15:11:25.127 [DEBUG] [sun.net.www.protocol.http.HttpURLConnection] sun.net.www.MessageHeader@52f17bac10 pairs: {POST /androidpublisher/v3/applications/com.brobizz.selfservice.dev/edits HTTP/1.1: null}{Accept-Encoding: gzip}{Authorization: Bearer ......}{User-Agent: gradle-play-publisher Google-API-Java-Client Google-HTTP-Java-Client/1.23.0 (gzip)}{Content-Encoding: gzip}{Host: www.googleapis.com}{Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2}{Connection: keep-alive}{Content-type: application/x-www-form-urlencoded}{Content-Length: 20}
15:11:25.291 [DEBUG] [sun.net.www.protocol.http.HttpURLConnection] sun.net.www.MessageHeader@5ee2b2bd14 pairs: {null: HTTP/1.1 404 Not Found}{Vary: Origin}{Vary: X-Origin}{Content-Type: application/json; charset=UTF-8}{Content-Encoding: gzip}{Date: Wed, 17 Oct 2018 13:11:27 GMT}{Expires: Wed, 17 Oct 2018 13:11:27 GMT}{Cache-Control: private, max-age=0
}{X-Content-Type-Options: nosniff}{X-Frame-Options: SAMEORIGIN}{X-XSS-Protection: 1; mode=block}{Server: GSE}{Alt-Svc: quic=":443"; ma=2592000; v="44,43,39,35"}{Transfer-Encoding: chunked}
15:11:25.292 [INFO] [com.google.api.client.http.HttpTransport] -------------- RESPONSE --------------
HTTP/1.1 404 Not Found
Transfer-Encoding: chunked
Alt-Svc: quic=":443"; ma=2592000; v="44,43,39,35"
Server: GSE
X-Content-Type-Options: nosniff
Date: Wed, 17 Oct 2018 13:11:27 GMT
X-Frame-Options: SAMEORIGIN
Cache-Control: private, max-age=0
Content-Encoding: gzip
Vary: X-Origin
Vary: Origin
Expires: Wed, 17 Oct 2018 13:11:27 GMT
X-XSS-Protection: 1; mode=block
Content-Type: application/json; charset=UTF-8

15:11:25.293 [INFO] [com.google.api.client.http.HttpTransport] Total: 326 bytes
15:11:25.293 [INFO] [com.google.api.client.http.HttpTransport] {
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "applicationNotFound",
    "message": "No application was found for the given package name.",
    "locationType": "parameter",
    "location": "packageName"
   }
  ],
  "code": 404,
  "message": "No application was found for the given package name."
 }
}

15:11:25.295 [DEBUG] [org.gradle.internal.progress.DefaultBuildOperationExecutor] Completing Build operation 'Execute process for :app:processDevReleaseMetadata'
15:11:25.295 [DEBUG] [org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter] Removed task artifact state for {} from context.
15:11:25.295 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Finished executing task ':app:processDevReleaseMetadata'
15:11:25.295 [DEBUG] [org.gradle.internal.progress.DefaultBuildOperationExecutor] Completing Build operation 'Task :app:processDevReleaseMetadata'
15:11:25.295 [null] [org.gradle.internal.progress.DefaultBuildOperationExecutor]
15:11:25.295 [DEBUG] [org.gradle.internal.progress.DefaultBuildOperationExecutor] Build operation 'Task :app:processDevReleaseMetadata' completed
15:11:25.295 [INFO] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] :app:processDevReleaseMetadata (Thread[Daemon worker Thread 9,5,main]) completed. Took 0.288 secs.
15:11:25.296 [DEBUG] [org.gradle.internal.work.DefaultWorkerLeaseService] Worker lease root.1.36 completed (1 worker(s) in use)
15:11:25.296 [DEBUG] [org.gradle.internal.resources.AbstractTrackedResourceLock] Daemon worker Thread 9: released lock on root.1.36
15:11:25.296 [DEBUG] [org.gradle.internal.resources.AbstractTrackedResourceLock] Daemon worker Thread 9: released lock on :
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Daemon worker Thread 9,5,main]] finished, busy: 17.014 secs, idle: 55.606 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':' Thread 6,5,main]] finished, busy: 0.0 secs, idle: 1 mins 12.62 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':' Thread 2,5,main]] finished, busy: 0.0 secs, idle: 1 mins 12.621 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':' Thread 5,5,main]] finished, busy: 0.0 secs, idle: 1 mins 12.62 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':' Thread 7,5,main]] finished, busy: 0.0 secs, idle: 1 mins 12.62 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':' Thread 3,5,main]] finished, busy: 55.295 secs, idle: 17.326 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':',5,main]] finished, busy: 0.0 secs, idle: 1 mins 12.621 secs
15:11:25.296 [DEBUG] [org.gradle.execution.taskgraph.DefaultTaskPlanExecutor] Task worker [Thread[Task worker for ':' Thread 4,5,main]] finished, busy: 0.0 secs, idle: 1 mins 12.621 secs
15:11:25.299 [DEBUG] [org.gradle.internal.progress.DefaultBuildOperationExecutor] Completing Build operation 'Run tasks'
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] FAILURE: Build failed with an exception.
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * What went wrong:
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Execution failed for task ':app:processDevReleaseMetadata'.
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] > No application found for the package name com.brobizz.selfservice.dev. The first version of your app must be uploaded via the Play Store console.
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Try:
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Run with --stacktrace option to get the stack trace.  Run with --scan to get full insights.
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]
15:11:25.301 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Get more help at https://help.gradle.org
15:11:25.302 [ERROR] [org.gradle.internal.buildevents.BuildResultLogger]
15:11:25.302 [ERROR] [org.gradle.internal.buildevents.BuildResultLogger] BUILD FAILED in 1m 16s

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/internal/Validation.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/internal/Validation.kt
@@ -2,6 +2,7 @@ package com.github.triplet.gradle.play.internal
 
 import com.android.builder.model.Version
 import com.github.triplet.gradle.play.PlayPublisherExtension
+import com.google.api.client.googleapis.json.GoogleJsonResponseException
 import org.gradle.util.GradleVersion
 
 private val MIN_GRADLE_VERSION: GradleVersion = GradleVersion.version("4.1")
@@ -48,3 +49,6 @@ internal fun PlayPublisherExtension.validate() {
         }
     }
 }
+
+internal infix fun GoogleJsonResponseException.has(error: String) =
+        details?.errors.orEmpty().any { it.reason == error }

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/ProcessPackageMetadata.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/ProcessPackageMetadata.kt
@@ -20,7 +20,7 @@ open class ProcessPackageMetadata : PlayPublishTaskBase() {
         progressLogger.completed()
     }
 
-    private fun processVersionCodes() = read { editId ->
+    private fun processVersionCodes() = read(true) { editId ->
         progressLogger.progress("Downloading active version codes")
         val maxVersionCode = tracks().list(variant.applicationId, editId).execute().tracks
                 ?.map { it.releases ?: emptyList() }?.flatten()

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishApk.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishApk.kt
@@ -62,7 +62,7 @@ open class PublishApk : PlayPublishPackageBase() {
                     .trackUploadProgress(progressLogger, "APK")
                     .execute()
         } catch (e: GoogleJsonResponseException) {
-            return e.handleUploadFailures(content.file)
+            return handleUploadFailures(e, content.file)
         }
 
         handlePackageDetails(editId, apk.versionCode)

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishBundle.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishBundle.kt
@@ -59,7 +59,7 @@ open class PublishBundle : PlayPublishPackageBase() {
                     .trackUploadProgress(progressLogger, "App Bundle")
                     .execute()
         } catch (e: GoogleJsonResponseException) {
-            return e.handleUploadFailures(content.file)
+            return handleUploadFailures(e, content.file)
         }
 
         handlePackageDetails(editId, bundle.versionCode)

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishListing.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/PublishListing.kt
@@ -6,6 +6,7 @@ import com.github.triplet.gradle.play.internal.ImageType
 import com.github.triplet.gradle.play.internal.LISTINGS_PATH
 import com.github.triplet.gradle.play.internal.ListingDetail
 import com.github.triplet.gradle.play.internal.climbUpTo
+import com.github.triplet.gradle.play.internal.has
 import com.github.triplet.gradle.play.internal.isDirectChildOf
 import com.github.triplet.gradle.play.internal.orNull
 import com.github.triplet.gradle.play.internal.playPath
@@ -123,7 +124,7 @@ open class PublishListing : PlayPublishTaskBase() {
         try {
             listings().update(variant.applicationId, editId, locale, listing).execute()
         } catch (e: GoogleJsonResponseException) {
-            if (e.details?.errors.orEmpty().any { it.reason == "unsupportedListingLanguage" }) {
+            if (e has "unsupportedListingLanguage") {
                 // Rethrow for clarity
                 throw IllegalArgumentException("Unsupported locale $locale", e)
             } else {

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishPackageBase.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishPackageBase.kt
@@ -6,6 +6,7 @@ import com.github.triplet.gradle.play.internal.RELEASE_NOTES_MAX_LENGTH
 import com.github.triplet.gradle.play.internal.RELEASE_NOTES_PATH
 import com.github.triplet.gradle.play.internal.ReleaseStatus
 import com.github.triplet.gradle.play.internal.ResolutionStrategy
+import com.github.triplet.gradle.play.internal.has
 import com.github.triplet.gradle.play.internal.orNull
 import com.github.triplet.gradle.play.internal.readProcessed
 import com.github.triplet.gradle.play.internal.trackUploadProgress
@@ -93,27 +94,25 @@ abstract class PlayPublishPackageBase : PlayPublishTaskBase() {
         return this
     }
 
-    protected fun GoogleJsonResponseException.handleUploadFailures(file: File): Nothing? {
-        val isConflict = details?.errors.orEmpty().all {
-            it.reason == "apkUpgradeVersionConflict" || it.reason == "apkNoUpgradePath"
-        }
+    protected fun handleUploadFailures(e: GoogleJsonResponseException, file: File): Nothing? {
+        val isConflict = e has "apkUpgradeVersionConflict" || e has "apkNoUpgradePath"
         if (isConflict) {
             when (extension._resolutionStrategy) {
                 ResolutionStrategy.AUTO -> throw IllegalStateException(
                         "Concurrent uploads for variant ${variant.name}. Make sure to " +
                                 "synchronously upload your APKs such that they don't conflict.",
-                        this
+                        e
                 )
                 ResolutionStrategy.FAIL -> throw IllegalStateException(
                         "Version code ${variant.versionCode} is too low for variant ${variant.name}.",
-                        this
+                        e
                 )
                 ResolutionStrategy.IGNORE -> logger.warn(
                         "Ignoring APK ($file) for version code ${variant.versionCode}")
             }
             return null
         } else {
-            throw this
+            throw e
         }
     }
 

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishTaskBase.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishTaskBase.kt
@@ -4,6 +4,7 @@ import com.android.build.gradle.api.ApplicationVariant
 import com.github.triplet.gradle.play.PlayPublisherExtension
 import com.github.triplet.gradle.play.internal.AccountConfig
 import com.github.triplet.gradle.play.internal.PLUGIN_NAME
+import com.github.triplet.gradle.play.internal.has
 import com.github.triplet.gradle.play.internal.transport
 import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
 import com.google.api.client.googleapis.json.GoogleJsonResponseException
@@ -64,7 +65,7 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
         val id = try {
             request.execute().id
         } catch (e: GoogleJsonResponseException) {
-            if (e.isApplicationNotFound) {
+            if (e has "applicationNotFound") {
                 if (skipIfNotFound) {
                     return
                 } else {
@@ -88,7 +89,4 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
         block(it)
         commit(variant.applicationId, it).execute()
     }
-
-    private val GoogleJsonResponseException.isApplicationNotFound
-        get() = details?.errors.orEmpty().any { it.reason == "applicationNotFound" }
 }

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishTaskBase.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishTaskBase.kt
@@ -54,19 +54,23 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
         }.setApplicationName(PLUGIN_NAME).build()
     }
 
-    protected fun read(block: AndroidPublisher.Edits.(editId: String) -> Unit) {
+    protected fun read(skipIfNotFound: Boolean = false, block: AndroidPublisher.Edits.(editId: String) -> Unit) {
         val edits = publisher.edits()
         val request = edits.insert(variant.applicationId, null)
 
         val id = try {
             request.execute().id
         } catch (e: GoogleJsonResponseException) {
-            // Rethrow for clarity
-            if (e.details?.errors.orEmpty().any { it.reason == "applicationNotFound" }) {
-                throw IllegalArgumentException(
-                        "No application found for the package name ${variant.applicationId}. " +
-                                "The first version of your app must be uploaded via the " +
-                                "Play Store console.", e)
+            if (e.isApplicationNotFound) {
+                if (skipIfNotFound) {
+                    null
+                } else {
+                    // Rethrow for clarity
+                    throw IllegalArgumentException(
+                            "No application found for the package name ${variant.applicationId}. " +
+                                    "The first version of your app must be uploaded via the " +
+                                    "Play Store console.", e)
+                }
             } else if (e.statusCode == 401) {
                 throw IllegalArgumentException("Invalid service account credentials.", e)
             } else {
@@ -74,11 +78,17 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
             }
         }
 
-        edits.block(id)
+        // If application wasn't found we might want to skip calling the block altogether
+        if (id != null) {
+            edits.block(id)
+        }
     }
 
     protected fun write(block: AndroidPublisher.Edits.(editId: String) -> Unit) = read {
         block(it)
         commit(variant.applicationId, it).execute()
     }
+
+    private val GoogleJsonResponseException.isApplicationNotFound
+        get() = details?.errors.orEmpty().any { it.reason == "applicationNotFound" }
 }

--- a/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishTaskBase.kt
+++ b/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/internal/PlayPublishTaskBase.kt
@@ -54,7 +54,10 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
         }.setApplicationName(PLUGIN_NAME).build()
     }
 
-    protected fun read(skipIfNotFound: Boolean = false, block: AndroidPublisher.Edits.(editId: String) -> Unit) {
+    protected fun read(
+            skipIfNotFound: Boolean = false,
+            block: AndroidPublisher.Edits.(editId: String) -> Unit
+    ) {
         val edits = publisher.edits()
         val request = edits.insert(variant.applicationId, null)
 
@@ -63,7 +66,7 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
         } catch (e: GoogleJsonResponseException) {
             if (e.isApplicationNotFound) {
                 if (skipIfNotFound) {
-                    null
+                    return
                 } else {
                     // Rethrow for clarity
                     throw IllegalArgumentException(
@@ -78,10 +81,7 @@ abstract class PlayPublishTaskBase : DefaultTask(), ExtensionOptions {
             }
         }
 
-        // If application wasn't found we might want to skip calling the block altogether
-        if (id != null) {
-            edits.block(id)
-        }
+        edits.block(id)
     }
 
     protected fun write(block: AndroidPublisher.Edits.(editId: String) -> Unit) = read {

